storage/libradb/src/pruner/mod.rs

```diff
@@ -220,43 +238,119 @@ impl Worker {
             }
         }
     }
+
+    /// Purge the stale node index so that after restart not too much already pruned stuff is dealt
+    /// with again (although no harm is done deleting those then non-existent things.)
+    ///
+    /// We issue (range) deletes on the index only periodically instead of after every pruning batch
+    /// to avoid sending too many deletions to the DB, which takes disk space and slows it down.
+    fn maybe_purge_index(&mut self) -> Result<()> {
+        const MIN_INTERVAL: Duration = Duration::from_secs(60);
+        const MIN_VERSIONS: u64 = 60000;
+
+        // A deletion is issued at most once in one minute and when the pruner has progressed by at
+        // least 60000 versions (assuming the pruner deletes as slow as 1000 versions per second,
+        // this imposes at most one minute of work in vain after restarting.)
+        let now = Instant::now();
+        if now - self.index_purged_at > MIN_INTERVAL {
+            let least_readable_version = self.least_readable_version.load(Ordering::Relaxed);
+
+            if least_readable_version - self.index_purged_till > MIN_VERSIONS {
+                self.db.range_delete::<StaleNodeIndexSchema, Version>(
+                    &self.index_purged_till,
```

Comment:
Will RocksDB notice that the current and previous deleted key range are adjacent (and hence can be combined) when creating tombstones in the memtable?  If Rocks is not that clever, maybe it would be better to explicitly force this by deleting from self.index_purged_till - 1.  I actually don't think it can be that clever since it has no way to know keys `00000001` and `00000002` are not straddling `000000011`

Reply:
I see I am deleting [a, b+1) in one batch and [b, c+1) in another makes the ranges overlab with each other. Fixing.

n.b
Physically, instead of combining them, it "fragment"s it: https://rocksdb.org/blog/2018/11/21/delete-range.html (That's why I don't always delete [0, least_readable_version + 1)). (But it doesn't affect your argument.)